### DISCUSSION TITLE
posts/Go/concurrence/

### DISCUSSION THREAD

**Discussion by Q1mi** [Q1mi/BlogComments#360] [Announcements]
https://www.liwenzhou.com/posts/Go/14_concurrence/ 

 李文周的Blog Go语言并发 goroutine channel go 线程 进程 锁 互斥锁 读写锁

**Comment by 272909106** [comment]
#### 题一
```golang
package main

import (
	"fmt"
	"math/rand"
	"strconv"
	"sync"
	"time"
)

var wg sync.WaitGroup
var once sync.Once
var lock sync.Mutex

func getRand(jobChan chan<- int64) {
	defer wg.Done()
	for i := 0; i < 24; i++ {
		// rand1 := rand.New(rand.NewSource(time.Now().UTC().UnixNano()))
		var randNum int64 = rand.Int63()
		fmt.Println(randNum)
		jobChan <- randNum
	}
	close(jobChan)
}

func putResult(jobChan <-chan int64, results chan<- int64) {
	defer wg.Done()
	for {
		i, ok := <-jobChan
		if !ok {
			break
		}
		int64Num := i
		lenNum := len(strconv.FormatInt(int64Num, 10))
		var sum int64 = 0
		for i := 1; i <= lenNum; i++ {
			last := int64Num % 10
			sum = sum + last
			int64Num = int64Num / 10
		}
		results <- sum
	}
	// once.Do(func() { close(results) })

}

func main() {
	start := time.Now()
	jobChan := make(chan int64, 100)
	results := make(chan int64, 100)

	wg.Add(1)
	go getRand(jobChan)

	for i := 1; i <= 24; i++ {
		wg.Add(1)
		go putResult(jobChan, results)
	}
	wg.Wait()
	close(results)
	for i := range results {
		fmt.Println(i)
	}
	end := time.Now()
	fmt.Println(end.Sub(start))
}
```

**Comment by jacobwwang** [comment]
2 个 goroutine 执行 f2 函数，这两个 goroutine 假设为 go1、go2。

会存在这种情况：
当 ch1 里只剩下最后一个元素时，go1 拿到最后一个元素，准备执行 x*x 操作，这时让出了时间片。
go2 在一个时钟周期内，关闭了 close(ch2)。
这时 go1 会往一个已关闭的 ch2 里 send 值。

**Comment by zym434597594** [comment]
使用sync.once关闭通道在发送端并发的情况还是会出问题的，它只能解决重复关闭的问题，但是无法解决其中一个关闭了，但是其他可能还在执行发送操作，同样会引发panic的

**Comment by GGGIsGGG** [comment]
比较下互斥锁和原子操作的性能时，我将i改为1000000之后，互斥锁的性能反而比原子操作好了。到底为啥呢？

**Comment by Q1mi** [comment]
> @GGGIsGGG
> 比较下互斥锁和原子操作的性能时，我将i改为1000000之后，互斥锁的性能反而比原子操作好了。到底为啥呢？

原来的示例代码有问题，已经更新了。

**Comment by Q1mi** [comment]
> @zym434597594
> 使用sync.once关闭通道在发送端并发的情况还是会出问题的，它只能解决重复关闭的问题，但是无法解决其中一个关闭了，但是其他可能还在执行发送操作，同样会引发panic的

原来的示例代码有问题，已经更换成一个单例模式的示例了。

**Comment by ourlang** [comment]
```go
package main

import (
	"fmt"
	"math/rand"
	"sync"
	"time"
)

var wg sync.WaitGroup

//计算一个64位随机数的各位的和
func randNumber(x int64) int64 {
	var sum int64 = 0
	for x > 0 {
		a := x % 10
		x = x / 10
		sum += a
	}
	return sum
}
// 生成int64的随机数放入通道ch1中
func createRand(ch1 chan<- int64) {
	for {
		int63 := rand.Int63()
		ch1 <- int63
		time.Sleep(1)
	}
}
//从通道ch1读取数据，然后计算各个位数之和存入ch2中
func readRand(ch1 <-chan int64, ch2 chan<- int64) {
	for  {
		value:=<-ch1
		number := randNumber(value)
		ch2<-number
		fmt.Println(value,number)
	}
}
func main() {
	var jobChan = make(chan int64, 100)
	var resultChan = make(chan int64, 100)
	wg.Add(25)
	go createRand(jobChan)

	for i := 0; i < 24; i++ {
		go readRand(jobChan,resultChan)
	}
	//循环打印数随机生成树的各位之和
	for value := range resultChan {
		fmt.Println(value)
	}
	wg.Wait()

}
```

**Comment by Bit0r** [comment]
使用无缓冲通道，则不能使用sync.Wait()等待。因为如果进行等待，则有一个通道既无法读，也无法写
```Go
package main

import (
	"fmt"
	"math/rand"
	"time"
)

func generate(jobChan chan<- int) {
	r := rand.New(rand.NewSource(time.Now().UnixNano()))
	for i := 0; i < 24; i++ {
		jobChan <- r.Int()
	}
	close(jobChan)
}

func calc(jobChan <-chan int, resultChan chan<- int) {
	result, x := 0, 0
	num := <-jobChan
	for num != 0 {
		x = num % 10
		num /= 10
		result += x
	}
	resultChan <- result
}

func main() {
	jobChan, resultChan := make(chan int), make(chan int)
	for i := 0; i < 24; i++ {
		go calc(jobChan, resultChan)
	}
	go generate(jobChan)
	for i := 0; i < 24; i++ {
		result := <-resultChan
		fmt.Println(result)
	}
	close(resultChan)
}

```

**Comment by yusonghu** [comment]
**写得很好，之前的那些锁弄懂了**

**Comment by miniyk2012** [comment]
目前写文件的Logger并发写时, 可能创建2次文件, 也可能导致文件被覆盖, 以后可以考虑加锁.
```golang
package main

import (
	"fmt"
	"io"
	"math/rand"
	"os"
	"sync"
)

const NUM = 100
const WORKER = 5

type Data struct {
	num     int64
	total   int64
	routine int
}

func (d Data) String() string {
	return fmt.Sprintf("sum(%d)=%d, worker=%d", d.num, d.total, d.routine)
}

func generate(job chan<- int64) {
	for i := 0; i < NUM; i++ {
		job <- rand.Int63n(1000)
	}
	close(job)
}

func calculate(job <-chan int64, result chan<- Data, jobNo int) {
	for num := range job {
		total := calNumSum(num)
		result <- Data{num, total, jobNo}
	}
}

func printResults(result <-chan Data) {
	for i := 0; i < NUM; i++ {
		data := <-result
		fmt.Printf("%d: %v\n", i, data)
	}
}

func calNumSum(num int64) int64 {
	var ret int64
	for num > 0 {
		ret += num % 10
		num = num / 10
	}
	return ret
}

func homework1() {
	jobChan := make(chan int64)
	resultChan := make(chan Data)
	for i := 0; i < WORKER; i++ {
		go calculate(jobChan, resultChan, i)
	}
	go generate(jobChan)

	printResults(resultChan)
}

// 使用接口的方式实现一个既可以往终端写日志也可以往文件写日志的简易日志库。
type Logger interface {
	Info(string)
	Close()
}

type FileLogger struct {
	filename string
	wg       sync.WaitGroup
}

func newFileLogger(filename string) *FileLogger {
	instance := &FileLogger{filename: filename}
	return instance
}

func (fl *FileLogger) Close() {
	fl.wg.Wait()
	fmt.Println("fileLogger closed")
}


func (fl *FileLogger) Info(msg string) {
	fl.wg.Add(1)
	go func() {
		defer fl.wg.Done()
		var f *os.File
		var err1 error
		if checkFileIsExist(fl.filename) { //如果文件存在
			f, err1 = os.OpenFile(fl.filename, os.O_APPEND|os.O_WRONLY, 0666) //打开文件
			fmt.Println("文件存在")
		} else {
			f, err1 = os.Create(fl.filename) //创建文件
			fmt.Println("文件不存在")
		}
		defer f.Close()
		n, err1 := io.WriteString(f, msg+"\n") //写入文件(字符串)
		if err1 != nil {
			panic(err1)
		}
		fmt.Printf("写入 %d 个字节\n", n)
	}()
}

func checkFileIsExist(filename string) bool {
	if _, err := os.Stat(filename); os.IsNotExist(err) {
		return false
	}
	return true
}

func newConsoleLogger() *ConsoleLogger {
	return &ConsoleLogger{}
}

type ConsoleLogger struct {
	wg sync.WaitGroup
}

func (cl *ConsoleLogger) Close() {
	cl.wg.Wait()
	fmt.Println("ConsoleLogger closed")
}

func (cl *ConsoleLogger) Info(msg string) {
	cl.wg.Add(1)
	go func() {
		defer cl.wg.Done()
		fmt.Println(msg)
	}()
}

func homework2() {
	var logger Logger
	fileLogger := newFileLogger("log.txt")
	logger = fileLogger
	defer logger.Close()
	logger.Info("Hello")
	logger.Info("how are you")

	consoleLogger := newConsoleLogger()
	logger = consoleLogger
	defer logger.Close()
	logger.Info("Hello")
	logger.Info("how are you")
}

func main() {
	homework1()
	println()
	homework2()
}

```

**Comment by xmf1142987374** [comment]
package main

import (
	"fmt"
	"math/rand"
	"sync"
	"time"
)

var wg sync.WaitGroup
//计算int64各位数的和
func sum(num int64)int64{
	var sum int64 = 0
	for num>0{
		sum+=(num%10)
		num/=10
	}
	return sum
}
//随机生成int64放到jobchan通道
func createInt64( jobchan chan<- int64){
	defer wg.Done()
	for{
		num:=rand.Int63()
		select{
			case jobchan<-num:
		}
	}
	
}
//去除jobchan通道的值求和存放到resultchan
func setResult(jobchan <-chan int64,resultchan chan<- int64){
	for{
		defer wg.Done()
		var result int64
		result=<-jobchan
		resultchan <- sum(result)
		fmt.Println(result,sum(result))
		time.Sleep(time.Second*2)
	}
}
func main(){
	var jobchan = make(chan int64,100)
	var resultchan = make(chan int64,100)
	wg.Add(1)
	go createInt64(jobchan)
	for i:=0;i<24;i++{
		wg.Add(1)
		go setResult(jobchan,resultchan)
	}
	for value:=range resultchan{
		fmt.Println(value)
	}
	wg.Wait()
	close(jobchan)
	close(resultchan)

}

**Comment by xmf1142987374** [comment]
package mylog

import (
	"fmt"
	"os"
	"path"
	"time"
)

//FileLogger 结构体
type FileLogger struct {
	maxFileSize int64
	level       uint
	filePath    string
	fileName    string
	file        *os.File
	errFile     *os.File
	logMsg      chan *LogMsg
}

//LogMsg 日志信息结构体
type LogMsg struct {
	level    uint
	msg      string
	fileName string
	funcName string
	now      string
	line     int
}

var advice = make(chan struct{}, 1)//初始化长度为1通道，用来标志后面文件是否关闭

//Close file
func (f *FileLogger) Close() {
	f.file.Close()
	f.errFile.Close()
}

//initFile 初始化文件指针对象
func (f *FileLogger) initFile() error {
	filepath := path.Join(f.filePath, f.fileName)
	//普通日志文件
	file, err := os.OpenFile(filepath, os.O_APPEND|os.O_CREATE|os.O_RDWR, 0)
	if err != nil {
		return err
	}
	//错误级别上日志文件
	errFile, err := os.OpenFile(filepath+".err", os.O_APPEND|os.O_CREATE|os.O_RDWR, 0)
	if err != nil {
		fmt.Println("init file failed")
		return err
	}
	f.file = file
	f.errFile = errFile
	advice <- struct{}{} //发送到通道
	for i := 0; i < 5; i++ {
		go f.writeLog()
	}
	// go f.writeLog()
	return nil
}

//check fileSize ! if fileSize>=maxFileSize it be splited !
func (f *FileLogger) checkFile(file *os.File) *os.File {

	select {
	case <-advice:
		fileInfo, err := file.Stat()
		if err != nil {
			fmt.Println("checkFile() one error")
			return nil
		}
		if fileInfo.Size() >= f.maxFileSize {
			select {
			case advice <- struct{}{}:
			default:
			}
			fileObj, err := f.splitFile(file)
			if err != nil {
				fmt.Println("checkFile() two error")
				return nil
			}
			if file != nil {
				return fileObj
			}

		} else {
			select {
			case advice <- struct{}{}:
			default:
			}
		}
	default:

	}
	return file

}

//split logFile
func (f *FileLogger) splitFile(file *os.File) (*os.File, error) {
	nowTime := time.Now().Format("20060102030405000")
	fileInfo, err := file.Stat()
	if err != nil {
		return nil, err
	}
	oldFileName := path.Join(f.filePath, fileInfo.Name())
	newFileName := fmt.Sprintf("%v.back%v", oldFileName, nowTime)
	select {
	case <-advice:
		time.Sleep(time.Second * 3)
		file.Close()
		os.Rename(oldFileName, newFileName)
		newFile, err := os.OpenFile(oldFileName, os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0)
		if err != nil {
			return nil, err
		}
		select {
		case advice <- struct{}{}:
		default:
		}
		return newFile, nil
	default:
		return nil, nil
	}

}

//log write
func (f *FileLogger) log(level string, format string, a ...interface{}) {

	if !enable(parseLevel(level), f.level) {
		return
	}
	msg := fmt.Sprintf(format, a...)
	now, fileName, funcName, line := getInfo(3)
	logMessage := &LogMsg{
		level:    parseLevel(level),
		msg:      msg,
		fileName: fileName,
		funcName: funcName,
		now:      now,
		line:     line,
	}
	select {
	case f.logMsg <- logMessage:
	default:
	}

}

//写日志
func (f *FileLogger) writeLog() {
	for {

		time.Sleep(time.Microsecond)
		f.file = f.checkFile(f.file)
		select {
		case logMessage := <-f.logMsg:
			logInfo := fmt.Sprintf("[%v]:[%v]:[%v:%v]:[%v]msg:%v\n", logMessage.now, logMessage.level, logMessage.fileName, logMessage.funcName, logMessage.line, logMessage.msg)
			fmt.Fprintf(f.file, logInfo)
			if logMessage.level >= ERROR {
				f.errFile = f.checkFile(f.errFile)
				fmt.Fprintf(f.errFile, logInfo)
			}
		default:

		}

	}

}

//Debug level
func (f *FileLogger) Debug(format string, a ...interface{}) {
	f.log("DEBUG", format, a...)
}

//Trace level
func (f *FileLogger) Trace(format string, a ...interface{}) {
	f.log("TRACE", format, a...)
}

//Info level
func (f *FileLogger) Info(format string, a ...interface{}) {
	f.log("INFO", format, a...)
}

//Warning level
func (f *FileLogger) Warning(format string, a ...interface{}) {
	f.log("WARNING", format, a...)
}

//Error level
func (f *FileLogger) Error(format string, a ...interface{}) {
	f.log("ERROR", format, a...)
}

//Fatal level
func (f *FileLogger) Fatal(format string, a ...interface{}) {
	f.log("FATAL", format, a...)
}

**Comment by fengcunwu** [comment]
> @miniyk2012
> 目前写文件的Logger并发写时, 可能创建2次文件, 也可能导致文件被覆盖, 以后可以考虑加锁.
> ```golang
> package main
> 
> import (
> 	"fmt"
> 	"io"
> 	"math/rand"
> 	"os"
> 	"sync"
> )
> 
> const NUM = 100
> const WORKER = 5
> 
> type Data struct {
> 	num     int64
> 	total   int64
> 	routine int
> }
> 
> func (d Data) String() string {
> 	return fmt.Sprintf("sum(%d)=%d, worker=%d", d.num, d.total, d.routine)
> }
> 
> func generate(job chan<- int64) {
> 	for i := 0; i < NUM; i++ {
> 		job <- rand.Int63n(1000)
> 	}
> 	close(job)
> }
> 
> func calculate(job <-chan int64, result chan<- Data, jobNo int) {
> 	for num := range job {
> 		total := calNumSum(num)
> 		result <- Data{num, total, jobNo}
> 	}
> }
> 
> func printResults(result <-chan Data) {
> 	for i := 0; i < NUM; i++ {
> 		data := <-result
> 		fmt.Printf("%d: %v\n", i, data)
> 	}
> }
> 
> func calNumSum(num int64) int64 {
> 	var ret int64
> 	for num > 0 {
> 		ret += num % 10
> 		num = num / 10
> 	}
> 	return ret
> }
> 
> func homework1() {
> 	jobChan := make(chan int64)
> 	resultChan := make(chan Data)
> 	for i := 0; i < WORKER; i++ {
> 		go calculate(jobChan, resultChan, i)
> 	}
> 	go generate(jobChan)
> 
> 	printResults(resultChan)
> }
> 
> // 使用接口的方式实现一个既可以往终端写日志也可以往文件写日志的简易日志库。
> type Logger interface {
> 	Info(string)
> 	Close()
> }
> 
> type FileLogger struct {
> 	filename string
> 	wg       sync.WaitGroup
> }
> 
> func newFileLogger(filename string) *FileLogger {
> 	instance := &FileLogger{filename: filename}
> 	return instance
> }
> 
> func (fl *FileLogger) Close() {
> 	fl.wg.Wait()
> 	fmt.Println("fileLogger closed")
> }
> 
> 
> func (fl *FileLogger) Info(msg string) {
> 	fl.wg.Add(1)
> 	go func() {
> 		defer fl.wg.Done()
> 		var f *os.File
> 		var err1 error
> 		if checkFileIsExist(fl.filename) { //如果文件存在
> 			f, err1 = os.OpenFile(fl.filename, os.O_APPEND|os.O_WRONLY, 0666) //打开文件
> 			fmt.Println("文件存在")
> 		} else {
> 			f, err1 = os.Create(fl.filename) //创建文件
> 			fmt.Println("文件不存在")
> 		}
> 		defer f.Close()
> 		n, err1 := io.WriteString(f, msg+"\n") //写入文件(字符串)
> 		if err1 != nil {
> 			panic(err1)
> 		}
> 		fmt.Printf("写入 %d 个字节\n", n)
> 	}()
> }
> 
> func checkFileIsExist(filename string) bool {
> 	if _, err := os.Stat(filename); os.IsNotExist(err) {
> 		return false
> 	}
> 	return true
> }
> 
> func newConsoleLogger() *ConsoleLogger {
> 	return &ConsoleLogger{}
> }
> 
> type ConsoleLogger struct {
> 	wg sync.WaitGroup
> }
> 
> func (cl *ConsoleLogger) Close() {
> 	cl.wg.Wait()
> 	fmt.Println("ConsoleLogger closed")
> }
> 
> func (cl *ConsoleLogger) Info(msg string) {
> 	cl.wg.Add(1)
> 	go func() {
> 		defer cl.wg.Done()
> 		fmt.Println(msg)
> 	}()
> }
> 
> func homework2() {
> 	var logger Logger
> 	fileLogger := newFileLogger("log.txt")
> 	logger = fileLogger
> 	defer logger.Close()
> 	logger.Info("Hello")
> 	logger.Info("how are you")
> 
> 	consoleLogger := newConsoleLogger()
> 	logger = consoleLogger
> 	defer logger.Close()
> 	logger.Info("Hello")
> 	logger.Info("how are you")
> }
> 
> func main() {
> 	homework1()
> 	println()
> 	homework2()
> }
> 
> ```

您的第二个写一个文件日志写的有问题，我使用你的代码跑了一遍，发现调用两次info，在文件中有时候只能写进去一个，有时候会写两个，但是顺序不同，是您没有做好线程同步，在info函数中应该在最后加上wg.Wait()

**Comment by mogebingxue** [comment]
面试时问到了并发安全的map和waitgroup，后悔第一遍的时候没仔细看，只知道有这东西，用的时候api忘了，呜呜

**Comment by xiarong715** [comment]
设置 runtime.GOMAXPROCS(2) 时，仍然是执行完一个协程再执行第二个；是否因为main()也是一个协程，当设置runtime.GOMAXPROCS(3)时，两个协程会交错运行；

**Comment by newbigland** [comment]
1.7节缺少了sync包的Cond条件变量，这块内容也很重要，希望博主能补充上，谢谢！

**Comment by LIUXUCHONG** [comment]
G很好理解，就是个goroutine的，里面除了存放本goroutine信息外 还有与所在P的绑定等信息。
这句话是不是这么理解：
G很好理解，就是个goroutine队列，里面除了存放本goroutine信息外 还有与所在P的绑定等信息。

**Comment by LIUXUCHONG** [comment]
func hello() {
fmt.Println("Hello Goroutine!")
}
func main() {
go hello()
fmt.Println("main goroutine done!")
}
我这样运行结果偶尔是：
main goroutine done!
Hello Goroutine!

**Comment by Kyvrel** [comment]
```go
func main() {
	runtime.GOMAXPROCS(2)
	go a()
	go b()
	time.Sleep(time.Second)
}
```
线程数量设置成2 , 但是还是一个执行完, 才执行另外一个. 这是为什么啊
```shell
B: 1
B: 2
B: 3
B: 4
B: 5
B: 6
B: 7
B: 8
B: 9
A: 1
A: 2
A: 3
A: 4
A: 5
A: 6
A: 7
A: 8
A: 9

```

**Comment by SsnAgo** [comment]
佬，最后一个示例的CommonCounter相关的方法应该是用指针接受，不然count改不了，一直都是0

**Comment by PikaByter** [comment]
> ```go
> 	runtime.GOMAXPROCS(2)
> 	go a()
> 	go b()
> 	time.Sleep(time.Second)
> }
> ```

运行太快了，先进入输出的还是先完成，改成这样就能看出来了：
```go
func a() {
	for i := 1; i < 10; i++ {
		fmt.Println("A:", i)
		time.Sleep(1 * time.Second)
	}
}

func b() {
	for i := 1; i < 10; i++ {
		fmt.Println("B:", i)
		time.Sleep(1 * time.Second)
	}
}

func main() {
	runtime.GOMAXPROCS(2)
	go a()
	go b()
	time.Sleep(30 * time.Second)
}
```

**Comment by zhuzhuxia12** [comment]
//题目一

package main

import (
	"fmt"
	"math/rand"
)

var (
	jobChan    chan int64
	resultChan chan int64
)

//生成随机数
func getRand(jobChan chan<- int64) {
	for {
		jobChan <- rand.Int63()
	}
}

//计算结果并发送给resultChan
func getResult(jobChan <-chan int64, resultChan chan<- int64) {
	for {
		var randNum int64 = <-jobChan
		var sum int64 = 0
		for {
			sum += randNum % 10
			randNum = randNum / 10
			if randNum == 0 {
				break
			}
		}
		resultChan <- sum
	}
}

func main() {
	jobChan = make(chan int64)
	resultChan = make(chan int64)
	go getRand(jobChan)

	for i := 0; i < 24; i++ {
		go getResult(jobChan, resultChan)
	}

	for {
		var result int64 = <-resultChan
		fmt.Println(result)
	}
}

**Comment by QuYingying0628** [comment]
题1：
1. test包
```go
package test

import (
	"fmt"
	"math/rand"
	"sync"
	"time"
)

var WaitGroup sync.WaitGroup

// 向通道发送随机数
func ProduceRandomNumber(jobChan chan<- int64){
	rand.Seed(time.Now().UnixNano())
	for i:=0;i<300;i++{
		jobChan <- rand.Int63()
	}
	WaitGroup.Done()
}

// 计算随机数的各位数和
func CountSum(jobChan <-chan int64,resultChan chan<- int64)  {
	// 从通道拿值
	num:= <- jobChan
	// 初始各位数和
	sum:=int64(0)
	// 一个临时值代替通道拿到的值
	numTemp:=num
	for{
		sum+=numTemp%10
		numTemp/=10
		if numTemp==0{
			break
		}
	}
	// 向结果通道发送值
	resultChan <- sum
	fmt.Println(num,"各位数和为",sum)
	WaitGroup.Done()
}
```
2. main函数
```go
// 计算随机数的各位数之和
	jobChan:=make(chan int64,300)
	resultChan:=make(chan int64,300)

	// 向通道生成随机数
	test.WaitGroup.Add(1)
	go test.ProduceRandomNumber(jobChan)

	// 从通道读取数据并计算数据的各位数之和
	for i:=0;i<24;i++{
		test.WaitGroup.Add(1)
		go test.CountSum(jobChan,resultChan)
	}

	test.WaitGroup.Wait()
	close(resultChan)
	for sumResult := range resultChan{
		fmt.Println(sumResult)
	}
```

**Comment by beachant** [comment]
package main

import (
	"fmt"
	"math/rand"
	"runtime"
	"strconv"
	"strings"
	"sync"
)

var (
	wg sync.WaitGroup
)

func main() {
	wg.Add(1)
	jobChan := make(chan int64)
	go createRand(jobChan)

	resultChan := make(chan int64)
	for i := 0; i < 24; i++ {
		wg.Add(1)
		go calcAndSendResult(resultChan, jobChan)
	}
	go func() {
		wg.Wait()
		close(resultChan)
	}()
	for result := range resultChan {
		fmt.Printf("结果: %d\n", result)
	}
	fmt.Println("End.")
}

func createRand(jobChan chan<- int64) {
	defer wg.Done()
	for i := 0; i < 24; i++ {
		number := rand.Int63()
		jobChan <- number
		fmt.Printf("随机数%d: %d\n", i, number)
	}
	close(jobChan)
}

func calcAndSendResult(resultChan chan<- int64, jobChan <-chan int64) {
	defer wg.Done()
	for {
		num, ok := <-jobChan
		if ok {
			resultChan <- calc(num)
		}
	}
}

func calc(number int64) int64 {
	n := number
	var sum int64
	len := len(strconv.FormatInt(number, 10))
	for i := 1; i <= len; i++ {
		sum += number % 10
		number /= 10
	}
	fmt.Printf("goroutine: %d  number: %d  各位数之和: %d\n", getGID(), n, sum)
	return sum
}

func getGID() uint64 {
	var buf [64]byte
	n := runtime.Stack(buf[:], false)
	idField := string(buf[:n])
	fields := strings.Fields(idField)
	for i, f := range fields {
		if f == "goroutine" {
			id, _ := strconv.ParseUint(fields[i+1], 10, 64)
			return id
		}
	}
	return 0
}

**Comment by lihu3601456** [comment]
package main

import (
	"fmt"
	"math/rand"
	"sync"
	"time"
)

var wg sync.WaitGroup

func sync_main() {

	var jobChan chan int = make(chan int, 100)
	var resultChan chan int = make(chan int,100)

	wg.Add(24)

	for i := 0; i < 24; i++ {
		go calc_sum(jobChan, resultChan)
	}


	go get_rand64(jobChan, 24)

	//等待所有goroutine执行完毕
	wg.Wait()
	//关闭resultChan
	close(resultChan) 


	//循环读取resultChan中的结果
	for num := range resultChan{
		fmt.Printf("计算结果: %d\n", num)	
	}



}

// 开启一个 goroutine 循环生成int64类型的随机数，发送到jobChan
func get_rand64(jobcha chan int, index int) {

	for i := 0; i < index; i++ {

		// 初始化随机种子，保证每次运行生成不同随机数
		rand64_int := rand.New(rand.NewSource(time.Now().UnixNano())).Int63()
		fmt.Printf("获取64位随机数: %d\n", rand64_int)
		jobcha <- int(rand64_int)

	}
	close(jobcha)
}

// 从jobChan中取出随机数计算各位数的和，并将结果发送到resultChan
func calc_sum(jobChan chan int, resultChan chan int) {

	for num := range jobChan {

		sum := 0
		for num > 0 {
			a := num % 10
			num = num / 10
			sum += a
		}
		resultChan <- sum

		fmt.Println("循环结果: ",sum)
	}
		


	wg.Done()

}

**Comment by seabra2** [comment]
2.7 练习题
var wg sync.WaitGroup

func main() {
	for i := 0; i < 5; i++ {
		wg.Add(1)
		go func() {
			defer wg.Done()
			fmt.Println(i)
		}()
	}
	wg.Wait()
}
只有这样才能输出结果

**Comment by seabra2** [comment]
原子操作实例写的有问题，对于普通版使用的是值接收者，最后打印出来的结果一定就是0

**Comment by seabra2** [comment]
// Result 结构体用于存储计算结果
type Result struct {
	Num int64
	Sum int64
}

// 生产者，生成int64类型的随机数，发送到jobChan
func generateNumbers(jobChan chan<- int64) {
	for {
		num := rand.Int63()
		jobChan <- num
		// 为了方便观察，可以加一个短暂的延迟
		time.Sleep(time.Millisecond * 500)
	}
}

// 消费者（工人）：从 jobChan 取出数字，计算各位数之和，结果发往 resultChan
func worker(jobChan <-chan int64, resultChan chan<- *Result) {
	for n := range jobChan {
		sum := calculateSum(n)
		resultChan <- &Result{
			Num: n,
			Sum: sum,
		}
	}
}

// 计算各位数字之和的辅助函数
func calculateSum(n int64) int64 {
	var sum int64
	for n > 0 {
		sum += n % 10
		n /= 10
	}
	return sum
}

func main() {
	jobChan := make(chan int64, 100)
	resultChan := make(chan *Result, 100)
	go generateNumbers(jobChan)
	for i := 0; i < 24; i++ {
		go worker(jobChan, resultChan)
	}
	fmt.Println("开始计算随机数各位数之和...")
	for res := range resultChan {
		fmt.Printf("随机数: %d, 各位数之和: %d\n", res.Num, res.Sum)
	}
}
